### PR TITLE
[SWAP REFACTORING #2] refactor swap business logic

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary
https://docs.google.com/document/d/1QCuT2zT-G7fKB5azLmssj8O3yd2dwBpDuoRJuz0cw0U/edit#

## How to review
Sorry for the big PR, for easier code review, **please review per commit**, starting with the oldest

## Refactoring PR's list:
1. [Separate swap logic from view](https://github.com/cowprotocol/cowswap/pull/1040)
2. [Refactor swap business logic](https://github.com/cowprotocol/cowswap/pull/1039)
3. [Refactor wrap/unwrap flow](https://github.com/cowprotocol/cowswap/pull/1038)
4. New swap page
4.1. [Setup layout](https://github.com/cowprotocol/cowswap/pull/1043)
4.2. [Bind data rendering and logic](https://github.com/cowprotocol/cowswap/pull/1044)
4.3. [Recipient controls, trade rates, high slippage warning](https://github.com/cowprotocol/cowswap/pull/1045)
4.4. [Merge develop, code cleanup](https://github.com/cowprotocol/cowswap/pull/1046)
4.5. [Cosmos fixtures and documentation](https://github.com/cowprotocol/cowswap/pull/1041)

## Testing
Don't test this PR, test the final PR - [New swap page](https://github.com/cowprotocol/cowswap/pull/1041)

### CODE REVIEW

File: src/custom/pages/Swap/SwapMod.tsx [hunk 1]
```diff
@@ -490,7 +446,7 @@ export default function Swap({ history, location, className }: RouteComponentPro
                 showMaxButton={false}
                 hideBalance={false}
                 fiatValue={fiatValueOutput ?? undefined}
-                priceImpact={showWrap ? undefined : priceImpact}
+                priceImpact={onWrap ? undefined : priceImpact}
```
Comment (alfetopito) [inline src/custom/pages/Swap/SwapMod.tsx:449]: Already mentioned in the other PR, we should keep `showWrap` here, unless it has been removed.

**Review by alfetopito — APPROVED** [review]

File: src/custom/pages/Swap/hooks/useHandleSwap.ts [hunk 2]
```diff
@@ -74,10 +57,10 @@ export function useHandleSwap(input: HandleSwapInput): {
       const marketLabel = [trade?.inputAmount?.currency?.symbol, trade?.outputAmount?.currency?.symbol].join(',')
       swapAnalytics('Send', marketLabel)
 
-      setSwapState({ attemptingTxn: true, tradeToConfirm, showConfirm, swapErrorMessage: undefined, txHash: undefined })
+      sendTransaction(trade!)
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/custom/pages/Swap/hooks/useHandleSwap.ts]: Thanks!
Fixed: https://github.com/cowprotocol/cowswap/pull/1049/commits/5bbae947d1b75f6579ba79439938b14aeb13dd53

File: src/custom/pages/Swap/index.tsx [hunk 3]
```diff
@@ -3,50 +3,14 @@ import styled from 'styled-components/macro'
...
-export const ButtonLight = styled(ButtonLightMod)`
-  > div,
-  > div > div {
-    width: 100%;
-    word-break: break-all;
-    text-align: center;
-    align-items: center;
-  }
-`
-
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/custom/pages/Swap/index.tsx:85]: Yes, they are not used

File: src/custom/pages/Swap/swapFlow/index.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  logSwapFlow('STEP 1: confirm price impact')
+  if (input.context.priceImpact && !confirmPriceImpactWithoutFee(input.context.priceImpact)) return
+
+  logSwapFlow('STEP 2: send transaction')
+  swapFlowAnalytics.swap(input.swapFlowAnalyticsContext)
+  input.swapConfirmManager.sendTransaction(input.context.trade)
+
+  try {
+    logSwapFlow('STEP 3: wrap native token (optional)')
+    const wrapEth$ = input.flags.isSellEth ? wrapEthStep(input.wrapEthInput) : Promise.resolve()
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/custom/pages/Swap/swapFlow/index.ts:21]: Got rid of this in the next PRs.
p.s. $ at the end means this is a Promise instance

File: src/custom/pages/Swap/SwapMod.tsx (hunk 1 above)
Comment (shoom3301) [reply src/custom/pages/Swap/SwapMod.tsx:449]: Fixed in the next PRs

File: src/custom/pages/Swap/swapFlow/index.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import { signAndPostOrder } from 'utils/trade'
+import { presignOrderStep } from './steps/presignOrderStep'
+import { addPendingOrderStep } from 'pages/Swap/swapFlow/steps/addPendingOrderStep'
+import confirmPriceImpactWithoutFee from '@src/components/swap/confirmPriceImpactWithoutFee'
+import { getProviderErrorMessage, isRejectRequestProviderError } from 'utils/misc'
+import { logSwapFlow } from 'pages/Swap/swapFlow/logger'
+
+export async function swapFlow(input: SwapFlowContext) {
+  logSwapFlow('STEP 1: confirm price impact')
+  if (input.context.priceImpact && !confirmPriceImpactWithoutFee(input.context.priceImpact)) return
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/custom/pages/Swap/swapFlow/index.ts:13]: I also was thinking about using https://xstate.js.org for complex business logic.

File: src/custom/pages/Swap/SwapMod.tsx [hunk 6]
```diff
@@ -52,7 +52,7 @@ import { AMOUNT_PRECISION, INITIAL_ALLOWED_SLIPPAGE_PERCENT } from 'constants/in
 import FeeInformationTooltip from 'components/swap/FeeInformationTooltip'
 import { useWalletInfo } from 'hooks/useWalletInfo'
 import { useGetQuoteAndStatus } from 'state/price/hooks'
-import { SwapProps } from '.' // mod
+// mod
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/custom/pages/Swap/SwapMod.tsx]: Removed it in the next PR's :)

File: src/custom/pages/Swap/swapFlow/useSwapFlowContext.ts [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+function calculateValidTo(deadline: number): number {
+  // Need the timestamp in seconds
+  const now = Date.now() / 1000
+  // Must be an integer
+  const validTo = Math.floor(deadline + now)
+  // Should not be greater than what the contract supports
+  return Math.min(validTo, MAX_VALID_TO_EPOCH)
+}
+
+export function useSwapFlowContext(): SwapFlowContext | null {
```
Comment (anxolin) [inline src/custom/pages/Swap/swapFlow/useSwapFlowContext.ts:62]: this is kind of a big hook, although cool that we get all the swap context.

Probably we will find ways later on to keep simplifying and cleaning up. Maybe when we refactor further for limit orders

**Review by anxolin — APPROVED** [review]
Nice one

File: package.json [hunk 8]
```diff
@@ -303,6 +303,7 @@
     "workbox-core": "^6.1.0",
     "workbox-navigation-preload": "^6.1.0",
     "workbox-precaching": "^6.1.0",
-    "workbox-routing": "^6.1.0"
+    "workbox-routing": "^6.1.0",
+    "xstate": "^4.33.4"
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply package.json:302]: ooops, deleted: https://github.com/cowprotocol/cowswap/pull/1049/commits/422ab26753c9a161849ee990d364b185e2a1fdef

File: src/custom/pages/Swap/swapFlow/useSwapFlowContext.ts (hunk 7 above)
Comment (shoom3301) [reply src/custom/pages/Swap/swapFlow/useSwapFlowContext.ts:62]: For sure, this hooks is an intermediate solution

**Pull request merged**